### DISCUSSION TITLE
BL Touch Grid Centering - Take 2 & My Solution

### DISCUSSION THREAD

**Discussion by clintiepoo** [Jyers/Marlin#189] [Q&A] [answered]
Reference discussion here: https://github.com/Jyers/Marlin/discussions/60

I worked through a solution to get the BLTouch probing at the center of the board and to get a square-shaped, symmetrical grid.

### Center Probing at the middle of the board: Z_SAFE_HOMING
Remove offset from X and Y points. (configuration.h, lines 1495-1496)

Change this:
` #define Z_SAFE_HOMING_X_POINT ((X_BED_SIZE - 10) / 2)    // X point for Z homing`
`  #define Z_SAFE_HOMING_Y_POINT ((Y_BED_SIZE - 10) / 2)    // Y point for Z homing`

To this: 
` #define Z_SAFE_HOMING_X_POINT ((X_BED_SIZE) / 2)    // X point for Z homing`
` #define Z_SAFE_HOMING_Y_POINT ((Y_BED_SIZE) / 2)    // Y point for Z homing`

### Create automatic centered grid based on nozzle-to-probe position
PROBING_MARGIN, (configuration.h, ~line 1000) 

Change this:
`#define PROBING_MARGIN 10`

To this:
`constexpr float ntpo[] = NOZZLE_TO_PROBE_OFFSET;`
`#define PROBING_MARGIN max(abs(ntpo[0]),abs(ntpo[1]))`

My [BLTouch Mount](https://www.thingiverse.com/thing:4462870) has an offset of (-42, -10). This is equivalent to using a 42 mm margin, but I made it a variable so I don't have to change the same number in two places. 

@Jyers - curious if this is something you want to implement by default. This is essentially giving the biggest square that will fit on the board, taking into account the probe offset X/Y. I know very little about C++, so I had to use an intermediate variable to pull the X and Y values out. 

### Edit - I'm abandoning the square probe area in favor of more area for UBL.

**Comment by Jyers** [comment, marked as answer]
While this is certainly a good way to get your grid centered on the build plate, it is probably not something I'll implement by default. That is because while it looks nice to have the grid centered, there is no real benefit to it. In fact, by centering the grid through the increasing of the margin, you are making the probed area significantly smaller than if its allowed to utilize its full range to the left of the bed. The only real way around this is to increase the bed size and max positions to accommodate moving farther to the right, that way the probe can reach farther to the right allowing for a larger, centered grid.

**Reply by clintiepoo** [reply]
Very interesting. I understand about probing a smaller area, but I think the benefit is that having a centered area makes adjusting the bed by hand simpler. The probe points for me are close to the adjustment screws. Other than that, I agree that this is mostly for looks.

**Comment by swbog** [comment]
Couldn’t you increase xmax in probing settings to get it more into the right. Granted you have enough cable and track to run

**Comment by Woodcat64** [comment]
I'm not sure if it makes any real difference or if it has some negative effects, but was able to shift the mesh closer to center and right edge by

    #define X_MAX_POS (X_BED_SIZE + 20)
    #define PROBING_MARGIN_LEFT 25
    #define PROBING_MARGIN_RIGHT 25

**Reply by swbog** [reply]
This is what I was saying^ you keep your area on the left and add more probing area to the right. Problem is rigs with high x offsets. The more x offset the less bit of the right side of the bed you get because of the x max. Except I didn’t change my margins just increased x max

**Reply by clintiepoo** [reply]
I’ll give it a shot - thanks. Ideally I’d like to get the mesh directly over the adjustment screws

**Reply by clintiepoo** [reply]
I measured, and because of the chains I have installed on my printer wiring, I can only cheat X_MAX_POS by 6mm. I added that in. It's not exactly over the leveling screws, but it works for me.

**Reply by t3chguy** [reply]
I haven't fitted chains to X yet and trimmed one of the hotend carriage bolts to manage to get X_MAX_POS to 263mm, leaves a lot of flexibility for probing, especially as my probe offset is `-51.0, -6.5`

**Reply by Schnabulation** [reply]
With UBL you should even be able to fill in the point manually, that the probe can‘t reach...

**Reply by clintiepoo** [reply]
> With UBL you should even be able to fill in the point manually, that the probe can‘t reach...

I am curious about this. I’m using bilinear. Is it worth the hassle to switch?

**Reply by Schnabulation** [reply]
I don‘t know - I haven‘t done it yet but I want to. :-)
I‘ll let you know...

**Reply by clintiepoo** [reply]
With UBL, I am going to go back to doing something like I was before. An uncentered grid makes more sense to me if you're doing more than 4 points.

**Comment by Schnabulation** [comment]
Why make it complicated? I have set this and have my probing points absolutely centered (I also increased the bed size to 230x230 tho...)

`#define PROBING_MARGIN_LEFT 45
#define PROBING_MARGIN_RIGHT 0
#define PROBING_MARGIN_FRONT 10
#define PROBING_MARGIN_BACK 20`

**Reply by vthing** [reply]
I've been struggling with UBL and BLTouch for a few days now. One of the problems was that with _Autotilt Mesh_ it was always positioned at x 30.0 and so my BLTouch sensor was not above the print bed. As a result, _Autotilt Mesh_ always failed.

With the above settings in _configuration_adv.h_ it now works without problems.

Thanks!

Information related to my current hard- and software (Could be helpful for someone, maybe?):

- Printer: Creality Ender 3 v2
- Mainboard: Creality 32-Bit (4.2.2)
- Auto Leveling Sensor: ANTCLABS BLTouch 3.1 (connected to the 5-Pin Sensor Port on the MB)
- Filament Runout Sensor:  Creality 3D Ender 6/CR-6 (connected to the Filament Sensor Port on the MB)
- Display: Default Display that's comes with the Ender 3 v2

Firmware Branch: Jyers Marlin 2.0 Bugfix Branch
- Ender_3_V2_Extensible_UI (29.03.2021)
- UBL BLTouch configuration
- self compiled

**Reply by Jyers** [reply]
Although this may have solved your issue, if your probe was going off the bed during an operation, that generally points to incorrectly set x y probe offsets. These can be configured from the advanced menu and should be set to the distance between the probe and the nozzle in the x and y directions.

**Reply by vthing** [reply]
I had already thought of something similar and considered this only as a quick workaround. So far I had absolutely no idea where I can set this.


Since the whole 3D printing topic is completely new to me, I've only had my printer for 2 weeks, I'm still building up my knowledge by reading a lot in the Marlin documentation and various forums.

Thanks for the tip, I gratefully accept it and will try it out.

Basically I am for standardization and my adjustment in the source code is just a quick and dirty fix so I can finally try UBL. :-)

I'll have a look at the settings in the advanced menu and see if I can get it to work.

Thanks for your feedback. :-)

**Reply by vthing** [reply]
Just a quick feedback.

I undid the changes to the _configuration_adv.h_ and compiled the firmware and put it on the ender.
In the advanced menu I then adjusted my offset values for x/y and lo and behold, the sensor is always above the print bed.

Thanks for the hint. :-)

**Reply by Schnabulation** [reply]
> I then adjusted my offset values for x/y and lo and behold, the sensor is always above the print bed.

If you want to make this persistent over EEPROM resets you can set the X- and Y- offset in configuration.h under `Nozzle-to-Probe offsets` at around line 1021.

**Reply by gothcha** [reply]
Following from the discussions I have just run some tests on UBL using stock V2 hardware with BLTouch 3.1 and decided to try to set the mesh to the largest area that can be fully probed (my XY probe offset is -42,-11) I have based on the 10x10 UBL template & the X_MAX_POS travel limit is set to Jyers default of 249, this just clears the carriage screw.

I define a mesh to avoid the clips on back and front and use the max X travel by setting the following in ADV.H (X_BED_SIZE =230)

#if EITHER(MESH_BED_LEVELING, AUTO_BED_LEVELING_UBL)
  // Override the mesh area if the automatic (max) area is too large
  #define MESH_MIN_X 0
  #define MESH_MIN_Y 10
  #define MESH_MAX_X X_BED_SIZE - 23
  #define MESH_MAX_Y Y_BED_SIZE - 15
#endif

I then set #define PROBING_MARGIN 0         in configuration.h so it will probe all the mesh.

I then fix the 3 point tilt probe points Configuration_adv.h line 1796 .. instead of using the default calculated. Note these are the probe positions on the bed not the nozzle XY positions!

  #define PROBE_PT_1_X 20
  #define PROBE_PT_1_Y 21
  #define PROBE_PT_2_X 200
  #define PROBE_PT_2_Y 21
  #define PROBE_PT_3_X 115
  #define PROBE_PT_3_Y 209

This ensures that the tilt points are in a safe location.

Running the levelling now probes 100/100 points and covers most of the bed. I found I got a better set of data and no interpolation on the left which previously gave me higher offsets.

old mesh settings
![Old mesh](https://user-images.githubusercontent.com/80574048/113861300-f66e9000-979e-11eb-95c2-2eaee8efd752.jpeg)

new mesh
![New mesh](https://user-images.githubusercontent.com/80574048/113861335-01c1bb80-979f-11eb-825b-0d4dc5b977ee.jpeg)

Hope this is useful :)
